File: python/stereoICM.py
# Stereo Matching using Iterated Conditional Modes
# Computes a disparity map from a rectified stereo pair using Iterated Conditional Modes

import time
import numpy as np
import cv2 as cv
import matplotlib.pyplot as plt
from shiftArray import shiftArray

# Set parameters
dispLevels = 16 #disparity range: 0 to dispLevels-1
iterations = 60
lambda_ = 5 #weight of smoothness cost
trunc = 2 #truncation of smoothness cost

# Define data cost computation
dataCostComputation = lambda left,right: np.absolute(left-right) #absolute differences
#dataCostComputation = lambda left,right: (left-right)**2 #square differences

# Define smoothness cost computation
smoothnessCostComputation = lambda differences: lambda_*np.minimum(np.absolute(differences),trunc)

# Start timer
timerVal = time.time()

# Load left and right images in grayscale
leftImg = cv.imread("left.png",cv.IMREAD_GRAYSCALE)
rightImg = cv.imread("right.png",cv.IMREAD_GRAYSCALE)

# Apply a Gaussian filter
leftImg = cv.GaussianBlur(leftImg,(5,5),0.6)
rightImg = cv.GaussianBlur(rightImg,(5,5),0.6)

# Get the size
(rows,cols) = leftImg.shape

# Convert to int32
leftImg = leftImg.astype(np.int32)
rightImg = rightImg.astype(np.int32)

# Compute pixel-based matching cost (data cost)
dataCost = np.zeros((rows,cols,dispLevels),dtype=np.int32)
for d in range(dispLevels):
    #rightImgShifted = shiftArray(rightImg,[0,d])
    rightImgShifted = np.roll(rightImg,d,1) #less accurate, better performances
    dataCost[:,:,d] = dataCostComputation(leftImg,rightImgShifted)

# Initialize the disparity map
dispMap = np.argmin(dataCost,axis=2)

energy = np.zeros(iterations,dtype=np.int32)
d = np.arange(dispLevels)[np.newaxis,np.newaxis,:].astype(np.int32)

# Start iterations
for it in range(iterations):

    # Compute local energy
    localEnergy = dataCost + \
    smoothnessCostComputation(np.roll(dispMap,-1,0)[:,:,np.newaxis]-d) + \
    smoothnessCostComputation(np.roll(dispMap,-1,1)[:,:,np.newaxis]-d) + \
    smoothnessCostComputation(np.roll(dispMap,1,0)[:,:,np.newaxis]-d) + \
    smoothnessCostComputation(np.roll(dispMap,1,1)[:,:,np.newaxis]-d)
    
    # Compute the disparity map
    dispMap = np.argmin(localEnergy,axis=2)
    
    # Compute energy
    dataEnergy = np.sum(dataCost[np.arange(rows)[:,np.newaxis],np.arange(cols)[np.newaxis,:],dispMap])
    smoothnessEnergyHorizontal = np.sum(smoothnessCostComputation(np.diff(dispMap,n=1,axis=1)))
    smoothnessEnergyVertical = np.sum(smoothnessCostComputation(np.diff(dispMap,n=1,axis=0)))
    energy[it] = dataEnergy+smoothnessEnergyHorizontal+smoothnessEnergyVertical

    # Normalize the disparity map for display
    scaleFactor = 256/dispLevels
    dispImg = (dispMap*scaleFactor).astype(np.uint8)

    # Show disparity map
    plt.cla()
    plt.imshow(dispImg,cmap="gray")
    plt.show(block=False)
    plt.pause(0.01)

    # Show energy and iteration
    print("iteration: {0}/{1}, energy: {2}".format(it+1,iterations,energy[it]))

# Show convergence graph
plt.figure()
plt.plot(np.arange(1,iterations+1),energy,marker="o")
plt.xlabel("Iterations")
plt.ylabel("Energy")
plt.show(block=False)
plt.pause(0.01)

# Save disparity map
cv.imwrite("disparityICM.png",dispImg)

# Stop timer and display running time
elapsedTime = time.time()-timerVal
print("Running time: {:.2f} seconds".format(elapsedTime))

plt.show()

File: matlab/stereoICM.m
% Stereo Matching using Iterated Conditional Modes
% Computes a disparity map from a rectified stereo pair using Iterated Conditional Modes

% Set parameters
dispLevels = 16; %disparity range: 0 to dispLevels-1
iterations = 60;
lambda = 5; %weight of smoothness cost
trunc = 2; %truncation of smoothness cost

% Define data cost computation
dataCostComputation = @(left,right) abs(left-right); %absolute differences
%dataCostComputation = @(left,right) (left-right).^2; %square differences

% Define smoothness cost computation
smoothnessCostComputation = @(differences) lambda*min(abs(differences),trunc);

% Start timer
timerVal = tic();

% Load left and right images in grayscale
leftImg = rgb2gray(imread('left.png'));
rightImg = rgb2gray(imread('right.png'));

% Apply a Gaussian filter
leftImg = imgaussfilt(leftImg,0.6,'FilterSize',5);
rightImg = imgaussfilt(rightImg,0.6,'FilterSize',5);

% Get the size
[rows,cols] = size(leftImg);

% Convert to int32
leftImg = int32(leftImg);
rightImg = int32(rightImg);

% Compute pixel-based matching cost (data cost)
dataCost = zeros(rows,cols,dispLevels,'int32');
for d = 0:dispLevels-1
    %rightImgShifted = shiftArray(rightImg,[0,d]);
    rightImgShifted = circshift(rightImg,d,2); %less accurate, better performances
    dataCost(:,:,d+1) = dataCostComputation(leftImg,rightImgShifted);
end

% Initialize the disparity map
[~,ind] = min(dataCost,[],3);
dispMap = int32(ind-1);

figure
energy = zeros(iterations,1);
d = int32(permute(0:dispLevels-1,[1 3 2]));

% Start iterations
for it = 1:iterations

    % Compute local energy
    localEnergy = dataCost + ...
    smoothnessCostComputation(circshift(dispMap,-1,1)-d) + ...
    smoothnessCostComputation(circshift(dispMap,-1,2)-d) + ...
    smoothnessCostComputation(circshift(dispMap,1,1)-d) + ...
    smoothnessCostComputation(circshift(dispMap,1,2)-d);

    % Compute the disparity map
    [~,ind] = min(localEnergy,[],3);
    dispMap = int32(ind-1);
    
    % Compute energy
    [row,col] = ndgrid(1:size(ind,1),1:size(ind,2));
    linInd = sub2ind(size(dataCost),row,col,ind);
    dataEnergy = sum(sum(dataCost(linInd)));
    smoothnessEnergyHorizontal = sum(sum(smoothnessCostComputation(diff(dispMap,1,2))));
    smoothnessEnergyVertical = sum(sum(smoothnessCostComputation(diff(dispMap,1,1))));
    energy(it) = dataEnergy+smoothnessEnergyHorizontal+smoothnessEnergyVertical;
    
    % Normalize the disparity map for display
    scaleFactor = 256/dispLevels;
    dispImg = uint8(dispMap*scaleFactor);

    % Show disparity map
    imshow(dispImg)
    
    % Show energy and iteration
    fprintf('iteration: %d/%d, energy: %d\n',it,iterations,energy(it))
end

% Show convergence graph
figure
plot(1:iterations,energy,'bo-')
xlabel('Iterations')
ylabel('Energy')

% Save disparity map
imwrite(dispImg,'disparityICM.png')

% Stop timer and display running time
elapsedTime = toc(timerVal);
fprintf('Running time: %.2f seconds\n',elapsedTime)
